keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -940,13 +940,27 @@ set when it is enabled. Initially, they are declared as alpha. Even though they
 are alpha, changes to their schema are discouraged and would have to be done by
 using new field names.
 
-ResourceClaim, DeviceClass and ResourceClaimTemplate are new built-in types
-in `resource.k8s.io/v1alpha3`. This alpha group must be explicitly enabled in
+After promotion to beta they are still disabled by default unless the feature
+gate explicitly gets enabled. The feature gate remains off by default because
+DRA depends on a new API group which following the
+[convention](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/3136-beta-apis-off-by-default)
+is off by default.
```

Comment:
Shouldn't the feature gate be on by default, but just the API off by default? This is a significant distinction for managed k8s which may allow different controls for feature gates versus API enablement.

Reply:
The problem is that the code which gets enabled because of the feature gate then tries to use the API, which causes failures. For example, kube-scheduler fails to start because the ResourceSlice informer fails to sync the cache.

Reply:
But what's the advantage of enabling the feature gate when the API is off?

I can see how it might simplify the cluster administration, but that's all. It adds more code and thus additional complexity, too.

Reply:
And a need for more testing. "Feature gate on, API off" becomes another supported configuration that must work.

Reply:
In https://kubernetes.slack.com/archives/C0EG7JC6T/p1723820845568029 `@aojea` wrote:
>  I'm deadlocked here with the beta API to be always disabled requirement https://github.com/kubernetes/kubernetes/pull/126684, the feature gate depends on the APIs to work, so I don't have clear how to proceed https://github.com/kubernetes/kubernetes/pull/126684#discussion_r1718537405

That was about `MultiCIDRServiceAllocator`. It's beta in 1.31 with default `false`, using the same approach that I am proposing here:
https://github.com/kubernetes/kubernetes/blob/60efdae3d1c4f5ac7e3aa9acbfa8a818e1fb2093/pkg/features/versioned_kube_features.go#L486-L489

Reply:
Jordan confirmed it in https://github.com/kubernetes/kubernetes/pull/126684#discussion_r1719976789

Reply:
> Hmm. @deads2k a little help here?

Disabling beta APIs by default keeps us from growing an ecosystem of beta level serialization that is resistant to changes.  The beta featuregate being on means that a cluster-admin just has to enable the API to decide to us the feature.

Reply:
@pohly can you add a note about the validation Jordan recommends (require enabling the API if you enable the gate)?

Reply:
Sorry, what validation? Which comment from Jordan do you refer to?

"Require enabling the API if you enable the gate" is already how we document ["enabling DRA"](https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/#enabling-dynamic-resource-allocation). That won't change after promotion to beta. But I am not sure how to validate that.

Reply:
In Jordan's [comment](https://github.com/kubernetes/kubernetes/pull/126684#discussion_r1719976789) you linked above, Jordan suggests validating this. Perhaps there is an example in the code where this is done?

Reply:
> 3. validate user doesn't enable the gate and not the API

I'm not aware of any code which implements that. @aojea, you already had to deal with this in 1.31 for `MultiCIDRServiceAllocator`. The comment was about that feature. Did you implement such validation?

Reply:
I believe this is the last remaining PRR item, then I can approve. If you can add this validation as the plan-of-record, we can merge it now. If it turns out to be too difficult, we can skip it and update the KEP to reflect that.

Reply:
I've added one paragraph about this to the scheduler section, ending with "How to implement such a check reliably still needs to be determined."

I'm surprised that this is now required because I've not seen something like that before and a bit worried that this additional check itself will be a source of failures. I'll start a new discussion in #sig-architecture about this.

Reply:
As we are getting close to finalizing the KEP, I squashed. The latest change is https://github.com/kubernetes/enhancements/compare/012f94389f7f2263c34486d3a9697a2d040739d3..e64c56539f85dd5ae951bd0a51d564ce4ef5df02

Reply:
> > 3. validate user doesn't enable the gate and not the API
> 
> I'm not aware of any code which implements that. @aojea, you already had to deal with this in 1.31 for `MultiCIDRServiceAllocator`. The comment was about that feature. Did you implement such validation?

@pohly no, I didn't, enabling APIs in the multicidr feature without the feature gate is a noop, and enabling the feature without the APIs does not start the cluster at all. It seems that now is too late, but if I have to do it again I will implement the cross validation, seems the best option and probably we should document it for future cases